content/en/blog/_posts/2023-03-08-memory-qos-cgroups-v2/index.md

```diff
@@ -0,0 +1,137 @@
+---
+layout: blog
+title: 'Quality-of-Service for Memory Resources'
+date: 2021-11-26
+slug: qos-memory-resources
+---
+
+**Authors:** Tim Xu (Tencent Cloud)
+
+Kubernetes v1.27, released in April 2023, introduced changes to an alpha feature, Memory QoS, that improves how Linux nodes implement memory resource requests and limits. Support for Memory QoS was initially added in Kubernetes v1.22, and later some [limitations](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2570-memory-qos#reasons-for-changing-the-formula-of-memoryhigh-calculation-in-alpha-v127) around the formula for calculating memory.high were identified. Those limitations are addressed in Kubernetes v1.27.
+
+In prior releases, Kubernetes did not support memory quality guarantees. 
+For example, if you set container resources as follows:
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  name: example
+spec:
+  containers:
+  - name: nginx
+    resources:
+      requests:
+        memory: "64Mi"
+        cpu: "250m"
+      limits:
+        memory: "64Mi"
+        cpu: "500m"
+```
+`spec.containers[].resources.requests`(e.g. cpu, memory) is designed for scheduling. When you create a Pod, the Kubernetes scheduler selects a node for the Pod to run on. Each node has a maximum capacity for each of the resource types: the amount of CPU and memory it can provide for Pods. The scheduler ensures that, for each resource type, the sum of the resource requests of the scheduled Containers is less than the capacity of the node.
+
+`spec.containers[].resources.limits` is passed to the container runtime when the kubelet starts a container. CPU is considered a "compressible" resource. If your app starts hitting your CPU limits, Kubernetes starts throttling your container, giving your app potentially worse performance. However, it won’t be terminated. That is what "compressible" means.
+
+In cgroup v1, and prior to this feature, the container runtime never took into account and effectively ignored spec.containers[].resources.requests["memory"]. This is unlike CPU, in which the container runtime consider both requests and limits. Furthermore, memory actually can't be compressed in cgroup v1. Because there is no way to throttle memory usage, if a container goes past its memory limit it will be terminated by the kernel with an OOM (Out of Memory) kill.
```

Comment:
`spec.containers[*].resources.requests["memory"]` is ignored??? In what sense?

Reply:
kubelet will not take `requests` memory into account in cgroup v1 as there is no similar `memory.min` in cgroup v1.
The request of memory is mainly used in scheduling calculation that is based on node allocatable.

Reply:
Actually, this is content from https://kubernetes.io/blog/2021/11/26/qos-memory-resources/ that is not changed.